### ISSUE TITLE
Why move node_modules to parent directory?

### ISSUE THREAD

**Issue by nunovieira** [microsoft/vscode-docker#2071]
https://github.com/microsoft/vscode-docker/blob/f02f023ef2d42e88b0b6759fd75f9e6606545727/src/configureWorkspace/configureNode.ts#L28
This line intrigues me.
Why the `mv node_modules ../` part? Why move _node_modules_ up a directory?
Should this be in a generic Node.js Dockerfile template?

**Comment by bwateratmsft** [comment]
@philliphoff do you know why we do this?

**Comment by philliphoff** [comment]
It's from before my time on the extension, but my guess is that it was (at least an attempt at) a Docker image build layer optimization.  It may also have been done to allow the flexibility to bind-mount the local application source into the container (rather than always build within the container); by keeping the `node_modules` folder in the parent (in the container as well as locally), you no longer have to worry about platform-specific libraries on the local machine not matching the platform of the container (as described [here](https://www.docker.com/blog/keep-nodejs-rockin-in-docker/)).

**Comment by bwateratmsft** [comment]
I looked through Git history, it's been this way for a very long time.

@lostintangent more or less was the first to make it this way (up one level), albeit with changing to Yarn instead of NPM: https://github.com/microsoft/vscode-docker/pull/85
Then immediately changed it back to NPM: https://github.com/microsoft/vscode-docker/pull/86

Since then it's been moved a few times but the moving of `node_modules` up a level has persisted.

@lostintangent Do you remember the history behind this?

**Comment by lostintangent** [comment]
As @philliphoff mentioned, this was done to allow users to mount their local copy of their codebase, without that  copy needing to include the restored node_modules (which might also be built for an entirely different operating system). 

That said, it’s been quite a while, so I’m not sure if this strategy is still appropriate 😁

**Comment by bwateratmsft** [comment]
@nunovieira Does this arrangement affect one of your projects?

**Comment by nunovieira** [comment]
This doesn't affect me because I don't use this anymore.
I start the "dockerization" by getting the Dockerfile from the [Node.js guide](https://nodejs.org/en/docs/guides/nodejs-docker-webapp/). It's simpler and better explained.
I never move _node_modules_. For development, I bind mount the application directory and have an anonymous volume for _node_modules_. In a compose file, there's something like:
```
    volumes:
      - ./app:/usr/src/app:delegated
      - /usr/src/app/node_modules
```
To have intellisense in vscode, I copy the _node_modules_ from the container to the host. That code will never be executed in the host, it's just to get the definitions and introspection stuff. 

IMO, this template is to kick-start the initial development. It could have some comments to help first-timers get it to a production ready state.

**Comment by bwateratmsft** [comment]
Yes, we can look at ways to improve the default Node Dockerfile with additional comments. @uchenkadicode has done quite a bit of that for the Python Dockerfile.

**Comment by TheBrenny** [comment]
So according to [this blog post](https://www.docker.com/blog/keep-nodejs-rockin-in-docker/), it looks like it's so you can develop for two different systems at once (such as developing for a Windows host and a Linux container). But if you `.dockerignore` the `node_modules/` folder, then moving it up one won't matter anyway? Even if a `.dockeringore` isn't present, according to the docker reference, `COPY` only ["copies new files or directories"](https://docs.docker.com/engine/reference/builder/#:~:text=copies%20new%20files%20or%20directories). The word "new" stands out to me and makes it sound like if the file to be copied exists in the container, it won't be copied, meaning the entire `node_modules` directory would be ignored anyway?

I'm no docker expert, though. I was just curious as to why `&& mv node_modules ../` is present, and my hunting landed me here...

**Comment by grokpot** [comment]
The problem process is:
1. It's common to bind-mount your host workspace to your container workspace to allow hot-reloading (e.g., `- ./my-app:/app`)
2. The image is built, node modules are installed
3. Bind mount from Step 1 is attached, some/all dependencies are missing if the host does not contain them
4. A common workaround it to use a _data_ volume to essentially exclude `node_modules` from Step 1. (e.g., `- ./my-app/node_modules/`)
5. Data volumes (Step 4) are created once, meaning installing dependencies on host (maybe for the benefit of auto-writing to package.json) will not transfer to container.
6. Solutions for Step 5 are: 
    * Install packages directly on container => not fun
    * Always use `--renew-anon-volumes` when running docker up => slower, and [not currently possible to implement in the `docker up` command of vscode](https://github.com/microsoft/vscode-remote-release/issues/6298)
    * [Move `node_modules` up a folder](https://stackoverflow.com/a/66154057/2016473) which [takes advantage of the path resolver pattern](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) => common solution
7. Step 6 Solution 3 means modules used in the `scripts` of `package.json` installed via `node_modules` will not work. E.g., `nodemon server.js`
8. Solution to Step 7 is [to use `npx`](https://stackoverflow.com/a/38754980/2016473) which (I believe) uses the same `node_modules` resolver pattern as Step6 Solution 3

These steps result in a host-container environment with hot-reloading and detection/installation of dependency changes.

The irony of this is - when starting my new project:
1. I immediately commented out the line `RUN npm install --production --silent && mv node_modules ../` because I didn't understand _Why move node_modules to parent directory?_
2. I spent hours understanding the problem
3. In the end, I uncommented the line

I think y'all should leave it. It helps many people without them knowing the benefit.

**Comment by zq0904** [comment]
> The problem process is:
> 
> 1. It's common to bind-mount your host workspace to your container workspace to allow hot-reloading (e.g., `- ./my-app:/app`)
> 2. The image is built, node modules are installed
> 3. Bind mount from Step 1 is attached, but some/all dependencies are missing if the host `node_modules` differs from container's
> 4. A common workaround it to use a _data_ volume to exclude `node_modules` from Step 1. (e.g., `- ./my-app/node_modules/`). Although, a blank `node_modules` will still exist on the container.
> 5. Data volumes (Step 4) are created once, meaning installing dependencies on host (maybe for the benefit of auto-writing to package.json) will not transfer to container.
> 6. Solutions for Step 5 are:
>    
>    * Install packages directly on container => not fun
>    * Always use `--renew-anon-volumes` when running docker up => slower, and [not currently possible to implement in the `docker up` command of vscode](https://github.com/microsoft/vscode-remote-release/issues/6298) (Can we please reopen this?)
>    * [Move `node_modules` up a folder](https://stackoverflow.com/a/66154057/2016473) which [takes advantage of the path resolver pattern](https://nodejs.org/api/modules.html#modules_loading_from_node_modules_folders) => common solution
> 7. Step 6 Solution 3 means modules used in the `scripts` of `package.json` installed via `node_modules` will not work. E.g., `nodemon server.js`
> 8. Solution to Step 7 is [to use `npx`](https://stackoverflow.com/a/38754980/2016473) which (I believe) uses the same `node_modules` resolver pattern as Step 6 Solution 3
> 
> These steps result in a host-container environment with hot-reloading and detection/installation of dependency changes. Yay!
> 
> The irony of this is - when starting my new project:
> 
> 1. I immediately commented out the line `RUN npm install --production --silent && mv node_modules ../` because I didn't understand _Why move node_modules to parent directory?_
> 2. I spent hours understanding the problem
> 3. In the end, I uncommented the line
> 
> I think y'all should leave it. It helps many people without them knowing the benefit.
> 
> The only thing to change would be maybe `mv node_modules ../` to `mv node_modules /` so an absolute instead of relative path is used, which can trip up `package.json` execution in services like Google Cloud Run

First of all, thank you for your speech!

But I'm sorry, I still don't understand why, can you give me a simple example?
